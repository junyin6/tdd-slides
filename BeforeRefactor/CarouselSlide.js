import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

const Figure = styled.figure`
  margin: 0;
  padding: 0.5rem;
`;
const Img = styled.img`
  object-fit: cover;
  width: 100%;
  height: ${(props) =>
    typeof props.imgHeight === 'number'
      ? `${props.imgHeight}px`
      : props.imgHeight};
`;
const CarouselSlide = ({
  imgUrl,
  description,
  attribution,
  imgHeight,
  Img,
  ...rest
}) => {
  return (
    <Figure {...rest}>
      <Img src={imgUrl} imgHeight={imgHeight} />
      <figcaption>
        <strong>{description}</strong> {attribution}
      </figcaption>
    </Figure>
  );
};

CarouselSlide.propTypes = {
  Img: PropTypes.elementType,
  imgUrl: PropTypes.string.isRequired,
  description: PropTypes.node.isRequired,
  attribution: PropTypes.node,
  imgHeight: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
};

CarouselSlide.defaultProps = {
  imgHeight: 600,
  Img: Img,
};

export default CarouselSlide;
