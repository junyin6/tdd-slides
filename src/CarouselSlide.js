import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

const Img = styled.img`
  object-fit: cover;
  width: 100%;
  height: ${(props) =>
    typeof props.imgHeight === 'number'
      ? `${props.imgHeight}px`
      : props.imgHeight};
`;
const CarouselSlide = ({
  imgUrl,
  description,
  attribution,
  imgHeight,
  ...rest
}) => {
  return (
    <figure {...rest}>
      <Img src={imgUrl} imgHeight={imgHeight} />
      <figcaption>
        <strong>{description}</strong> {attribution}
      </figcaption>
    </figure>
  );
};

CarouselSlide.propTypes = {
  imgUrl: PropTypes.string.isRequired,
  description: PropTypes.node.isRequired,
  attribution: PropTypes.node,
  imgHeight: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
};

CarouselSlide.defaultProps = {
  imgHeight: 600,
};

export default CarouselSlide;
