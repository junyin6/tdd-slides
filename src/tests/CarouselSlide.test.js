import React from 'react';
import { shallow } from 'enzyme';
import CarouselSlide from '../CarouselSlide';

describe('CarouselSlide', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(
      <CarouselSlide
        imgURL="https://example.com/default.jpg"
        description="Default test image"
      />
    );
  });

  it('renders a <figure>', () => {
    expect(wrapper.type()).toBe('figure');
  });

  it('renders an <img> and <figcaption> as children', () => {
    expect(wrapper.childAt(0).type()).toBe('img');
    expect(wrapper.childAt(1).type()).toBe('figcaption');
  });

  it('passes "imgURL" to <img>', () => {
    const imgURL = 'https://example.com/image.png';
    wrapper.setProps({ imgURL });
    const img = wrapper.find('img');
    expect(img.prop('src')).toBe(imgURL);
  });

  it('passes "description" to <strong> in <figcaption>', () => {
    const description = 'what did I tell you';
    const attribution = 'Jun Yin';
    wrapper.setProps({ description, attribution });
    expect(wrapper.find('figcaption strong').text()).toBe(description);
    expect(wrapper.find('figcaption').text()).toBe(
      `${description} ${attribution}`
    );
  });

  it('passes other props as children to <figure>', () => {
    const onClick = () => {};
    const style = {};
    const className = 'carousel-slide';

    wrapper.setProps({ onClick, style, className });

    expect(wrapper.prop('onClick')).toBe(onClick);
    expect(wrapper.prop('style')).toBe(style);
    expect(wrapper.prop('className')).toBe(className);
  });
});
