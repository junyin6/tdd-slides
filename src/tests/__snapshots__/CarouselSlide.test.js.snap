// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CarouselSlide renders correctly 1`] = `
<CarouselSlide__Figure>
  <CarouselSlide__Img
    imgHeight={600}
    src="https://example.com/default.jpg"
  />
  <figcaption>
    <strong>
      description
    </strong>
     
    attribution
  </figcaption>
</CarouselSlide__Figure>
`;

exports[`Img renders correctly 1`] = `
.c0 {
  object-fit: cover;
  width: 100%;
  height: 500px;
}

<img
  className="c0"
  src="https//whatever.com/123"
/>
`;
